papers/Vvmax/calc_vvmax.py

```diff
@@ -0,0 +1,764 @@
+        #string = '{0:8} {1:1.3e} {2:6.4f} {3:6.4f}'.format(
+        #    int(FRB),JHz,Bzmax,Bdl)
+        
+        if write:
+            f.write(string),SNR[i],DM[i],DMG[i],freqMHz[i],tsamp[i],width[i],BeamThetaDeg[i]
+        
+        
+        #### constructs string to write to table in paper ####
+        if minz is not None:
+            string = '\\FRB{0:8} & {1:1.5f} & {2:1.2e} & {3:6.5f} & {4:1.2e} & {5:1.2e} & {6:6.3f} &'.format \
+            (names[i],cos.DL(zDM[i])/1e3,Enu[i],tempzmax,Volume,Vmax,Volume/Vmax)
+        else:
+            string = '\\FRB{0:8} & {1:1.3f} & {2:1.2e} & {3:6.3f} & {4:1.2e} & {5:1.2e} & {6:6.3f} &'.format \
+                (names[i],cos.DL(zDM[i])/1e3,Enu[i],tempzmax,Volume,Vmax,Volume/Vmax)
+        if laststrings is None:
+            tempstrings.append(string)
+            print(string)
+        else:
+            extra = ' {0:1.2e} & {1:6.3f} & {2:1.2e} & {3:1.2e} & {4:6.3f} \\'.format \
+            (Enu[i],tempzmax,Volume,Vmax,Volume/Vmax)
+            combined = laststrings[i]+extra
+            print(combined)
+    print("Calculation times are ",dt1,dt2)  
+    
+    if write:
+        f.close()
+    if laststrings is None:
+        return tempstrings
+    else:
+        return laststrings   
+            
+def integrate_volume(zmax,Nsfr=0.):
+    """
+    Integrates dV/dz from 0 to z
+    Weights according to SFR**n
+    """
+    result = quad(dVdtaudz,0,zmax,args=[Nsfr])
+    return result[0]
+    
+
+def dVdtaudz(z,Nsfr):
+    """
+    Volume element per dz per solid angle
+    """
+    weighted_dVdtau = cos.dVdtau(z) * SFR(z)**Nsfr[0]
+    return weighted_dVdtau
+    
+def SFR(z):
+    """
+    Madau & dickenson 2014
+    Arguments:
+        z (float): redshift
+    Leading constant: normalises to unity at z=0
+    """
+    return 1.0025738*(1+z)**2.7 / (1 + ((1+z)/2.9)**5.6)
+
+
+    
+def calc_effz(s0,DMcosmic0,z0,w0,DM0,freq,tsamp,alpha,ztrials,DMcosmicz):
+    """
+    calculates the relative SNR/SNRth that an FRB would have
+    as a function of redshift
+    
+    Includes luminosity distance and efficiency factors
+    """
+    # estimate DM at different z
+    DMz = DM0 + DMcosmicz - DMcosmic0
+    
+    # calculates the energy corresponding to 1 Jy ms at z0
+    E0 = F_to_E(1.,z0,alpha)
+    # calculates the fluence "per Jyms at z0" from an FRB at z
+    modFz = E_to_F(E0,ztrials,alpha)
+    
+    eff0 = zefficiency(z0,w0,DM0,freq,tsamp,z0,DM0)
+    effz = zefficiency(z0,w0,DM0,freq,tsamp,ztrials,DMz)
+    
+    # using Wayne's method
+    # this just scaled effective width as 1+z, hence
+    # efficiency went as (1+z)**-0.5
+    # i.e. it ignores the fact the intrinsic width behaves differently...
+    # eff0 = (1+z0)**-0.5
+    # effz = (1+ztrials)**-0.5
+    
+    sz = s0 * (effz/eff0) * modFz
+    
+    return sz,eff0,effz,modFz
+    
+
+def calc_zmax(s0,z0,DM0,DMcosmic0,w0,freq,tsamp,state,Mz,MDM,alpha):
```

Comment:
this looks like a very useful method for the community
   there is a somewhat similar one in the `FRB` repo already

Reply:
Yep, it's in the main code now, sort-of kind0of properly documented